src/transform/src/map_lifting.rs

```diff
@@ -149,15 +205,33 @@ impl LiteralLifting {
                 // reach a fixed point under the transformations.
                 let mut literals = self.action(input, gets);
                 if !literals.is_empty() {
+                    // Discard literals that are not projected
                     let input_arity = input.arity();
-                    if let Some(project_max) = outputs.iter().max() {
-                        // Discard literals that are not projected.
-                        // TODO(frank): this could also discard intermediate
-                        // literals that are not projected, with more thought.
-                        while *project_max + 1 < input_arity + literals.len()
-                            && !literals.is_empty()
-                        {
-                            literals.pop();
+                    let mut used_literals = vec![false; literals.len()];
+                    for i in 0..outputs.len() {
+                        if outputs[i] >= input_arity {
+                            used_literals[outputs[i] - input_arity] = true;
+                        }
+                    }
+                    if used_literals.iter().filter(|x| !*x).count() > 0 {
+                        let mut skipped_literals = 0;
+                        let mut new_literals = Vec::new();
+                        let mut literal_map = vec![0; literals.len()];
+                        for i in 0..literals.len() {
```

Comment:
For this and the other enum branches, we prefer using `Iterator` when rust reference borrowing rules allow to cut down on the number of variables/nested loops and make the code easier to understand. If you have to use a bunch of variables to do something like permutation, add a comment on the beginning to explain what all the variables do.

An example how iterator usage in this enum branch can reduce the number of variables.
```
// For each entry in `new_literals`, store 1) the column in `outputs` referencing the entry.
// and 2) the location of the copy of the entry in the old `literals`
let new_literal_spec = outputs.iter().enumerate().filter_map(|out_col, old_in_col| if old_in_col >= input_arity {
  Some((out_col, inp_col-input_arity))
} else {
  None
});
let new_literals = new_literal_spec.iter().map(|(_out_col, old_in_col)| literals[i].clone()).collect();
// Remap remaining projected literals
for (new_in_col, (out_col, _old_in_col)) in new_literal_spec.iter().enumerate() {
   outputs[out_col] = new_in_col + input_arity;
}
```

Reply:
Yeah, using too many variables is a red flag and makes the code hard to understand. The code you propose is more elegant and descriptive.

One little issue it introduces is the duplication of literals that are projected more than once. For example:

```build
(project
  (map (get x) [1 2 3])
  [#3 #3])
----
%0 =
| Get x (u0)
| Map 1, 2, 3
| Project (#3, #3)

build apply=LiteralLifting
(project
  (map (get x) [1 2 3])
  [#3 #3])
----
%0 =
| Get x (u0)
| Map 2, 2
| Project (#2, #3)
```

with the existing code in the branch the result is:

```
build apply=LiteralLifting
(project
  (map (get x) [1 2 3])
  [#3 #3])
----
%0 =
| Get x (u0)
| Map 2
| Project (#2, #2)
```

I'll try to see how to avoid the duplicates.

Reply:
The only way I can think of for removing duplicates is by using an intermediate map where the key is the position in the old `literals` and the value is the position in `new_literals`. The final code would look like this:

```
                    // Discard literals that are not projected
                    let input_arity = input.arity();
                    // For each entry in `new_literals`, store 1) the column
                    // in `outputs` referencing the entry. and 2) the
                    // location of the copy of the entry in the old `literals`
                    let new_literal_spec = outputs
                        .iter()
                        .enumerate()
                        .filter_map(|(out_col, old_in_col)| {
                            if *old_in_col >= input_arity {
                                Some((out_col, old_in_col - input_arity))
                            } else {
                                None
                            }
                        })
                        .collect::<Vec<(usize, usize)>>();
                    // Maps positions in `literals` with positions in
                    // new_literals` since `new_literal_spec` may contain
                    // several entries with the same `old_in_col` value.
                    let mut new_literal_map = HashMap::new();
                    for (_out_col, old_in_col) in new_literal_spec.iter() {
                        let new_in_col = new_literal_map.len();
                        new_literal_map.entry(old_in_col).or_insert(new_in_col);
                    }
                    if new_literal_map.len() != literals.len() {
                        let new_literals = new_literal_map
                            .iter()
                            .map(|(old_in_col, _out_cols)| literals[**old_in_col].clone())
                            .collect();
                        // Remap remaining projected literals
                        for (out_col, old_in_col) in new_literal_spec.iter() {
                            let new_in_col = new_literal_map.get(old_in_col).unwrap();
                            outputs[*out_col] = new_in_col + input_arity;
                        }
                        literals = new_literals;
                    }

```

Reply:
This is the best I've been able to write given my limited Rust skills to avoid adding duplicates:

```
                    let input_arity = input.arity();
                    // For each input literal contains a vector with the `output` positions
                    // that references it.
                    let used_literals = outputs
                        .iter()
                        .enumerate()
                        .filter(|(_, x)| **x >= input_arity)
                        .map(|(out_col, old_in_col)| (old_in_col - input_arity, out_col))
                        .into_group_map();
                    if used_literals.len() != literals.len() {
                        // Discard literals that are not projected
                        let new_literals = literals
                            .iter()
                            .enumerate()
                            .filter(|(old_in_col, _)| used_literals.contains_key(old_in_col))
                            .map(|(_, l)| l.clone())
                            .collect();
                        literals
                            .iter()
                            .enumerate()
                            .filter_map(|(old_in_col, _)| used_literals.get(&old_in_col))
                            .enumerate() // adds positions in `new_literals`
                            .for_each(|(new_in_col, out_cols)| {
                                // update the references to the literal in `output`
                                for out_col in out_cols {
                                    outputs[*out_col] = input_arity + new_in_col;
                                }
                            });
                        literals = new_literals;
                    }
```

Is it any better? Thanks!

Reply:
To avoid iterating over literals every time, you can take the contents out of the `HashMap` into a `Vec` and then sort the `Vec`.
```
                    let mut used_literals = outputs
                        .iter()
                        .enumerate()
                        .filter(|(_, x)| **x >= input_arity)
                        .map(|(out_col, old_in_col)| (old_in_col - input_arity, out_col))
                        .into_group_map() // eliminate duplicates
                        .drain().collect::<Vec<_>>(); 
                    // By putting the data into a Vec and sorting, we guarantee a reliable order.
                    used_literals.sort();
                    literals = used_literals().iter().map(|(old_in_col, _)| literals[old_in_col].clone()).collect::<Vec<_>>();
                    for (new_in_col, (_old_in_col, out_cols)) in new_literal_spec.iter().enumerate() {
                       for out_col in out_cols {
                          outputs[out_col] = new_in_col + input_arity;
                       }
                    }
```

On a side note: personally, I wonder if whether or not introducing duplicates of literals matter, especially once projection pushdown exists (#6449). I don't think it matters from a rendering performance perspective. It might matter from query planning/semantic rewrite perspective, but investigating that is outside the scope of this PR.

Reply:
Thank you. The code now looks like this:

```
                    let input_arity = input.arity();
                    // For each input literal contains a vector with the `output` positions
                    // that references it. By putting data into a Vec and sorting, we
                    // guarantee a reliable order.
                    let mut used_literals = outputs
                        .iter()
                        .enumerate()
                        .filter(|(_, x)| **x >= input_arity)
                        .map(|(out_col, old_in_col)| (old_in_col - input_arity, out_col))
                        // group them to avoid adding duplicated literals
                        .into_group_map()
                        .drain()
                        .collect::<Vec<_>>();

                    if used_literals.len() != literals.len() {
                        used_literals.sort();
                        // Discard literals that are not projected
                        literals = used_literals
                            .iter()
                            .map(|(old_in_col, _)| literals[*old_in_col].clone())
                            .collect::<Vec<_>>();
                        // Update the references to the literal in `output`
                        for (new_in_col, (_old_in_col, out_cols)) in
                            used_literals.iter().enumerate()
                        {
                            for out_col in out_cols {
                                outputs[*out_col] = input_arity + new_in_col;
                            }
                        }
                    }

```

Without knowing much about rendering or the actual execution of the query, I think it should be a rule of all semantic transforms like this one to always result in a more concise plan.